// @chakray/marked/esm5/marked/enroute.attr.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Input, Directive, ElementRef, HostListener } from '@angular/core';
import { Router } from '@angular/router';
var CmEnrouteAttr = /** @class */ (function () {
    function CmEnrouteAttr(el, router) {
        this.el = el;
        this.router = router;
        this.baseUrl = '';
    }
    /**
     * @param {?} event
     * @return {?}
     */
    CmEnrouteAttr.prototype.onClick = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        /** @type {?} */
        var attr = event.target.getAttribute('href') || '';
        /** @type {?} */
        var m = attr.split('#');
        /** @type {?} */
        var href = m[0];
        /** @type {?} */
        var fragment = m[1];
        if (href.startsWith('http')) {
            return;
        }
        if (href.includes('..')) {
            /** @type {?} */
            var meg = href.split('/');
            /** @type {?} */
            var n = meg.reduce((/**
             * @param {?} r
             * @param {?} k
             * @return {?}
             */
            function (r, k) {
                return k === '..' ? r + 1 : r;
            }), 0);
            /** @type {?} */
            var seg = this.baseUrl.split('/');
            seg = seg.slice(0, seg.length - n);
            meg.splice(0, n);
            href = seg.join('/') + '/' + meg.join('/');
        }
        else {
            href = this.baseUrl + '/' + href;
        }
        if (event.target.tagName === 'A') {
            if (href) {
                this.router.navigate([href], { fragment: fragment });
            }
            else {
                this.router.navigate([], { fragment: fragment });
            }
            event.preventDefault();
        }
        else {
            return;
        }
    };
    CmEnrouteAttr.decorators = [
        { type: Directive, args: [{
                    selector: '[cmEnroute]'
                },] }
    ];
    /** @nocollapse */
    CmEnrouteAttr.ctorParameters = function () { return [
        { type: ElementRef },
        { type: Router }
    ]; };
    CmEnrouteAttr.propDecorators = {
        baseUrl: [{ type: Input, args: ['cmEnroute',] }],
        onClick: [{ type: HostListener, args: ['click', ['$event'],] }]
    };
    return CmEnrouteAttr;
}());
export { CmEnrouteAttr };
if (false) {
    /** @type {?} */
    CmEnrouteAttr.prototype.baseUrl;
    /**
     * @type {?}
     * @private
     */
    CmEnrouteAttr.prototype.el;
    /**
     * @type {?}
     * @private
     */
    CmEnrouteAttr.prototype.router;
}
