// @chakray/marked/esm2015/marked/enroute.attr.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Input, Directive, ElementRef, HostListener } from '@angular/core';
import { Router } from '@angular/router';
export class CmEnrouteAttr {
    /**
     * @param {?} el
     * @param {?} router
     */
    constructor(el, router) {
        this.el = el;
        this.router = router;
        this.baseUrl = '';
    }
    /**
     * @param {?} event
     * @return {?}
     */
    onClick(event) {
        /** @type {?} */
        const attr = event.target.getAttribute('href') || '';
        /** @type {?} */
        const m = attr.split('#');
        /** @type {?} */
        let href = m[0];
        /** @type {?} */
        const fragment = m[1];
        if (href.startsWith('http')) {
            return;
        }
        if (href.includes('..')) {
            /** @type {?} */
            const meg = href.split('/');
            /** @type {?} */
            const n = meg.reduce((/**
             * @param {?} r
             * @param {?} k
             * @return {?}
             */
            (r, k) => {
                return k === '..' ? r + 1 : r;
            }), 0);
            /** @type {?} */
            let seg = this.baseUrl.split('/');
            seg = seg.slice(0, seg.length - n);
            meg.splice(0, n);
            href = seg.join('/') + '/' + meg.join('/');
        }
        else {
            href = this.baseUrl + '/' + href;
        }
        if (event.target.tagName === 'A') {
            if (href) {
                this.router.navigate([href], { fragment });
            }
            else {
                this.router.navigate([], { fragment });
            }
            event.preventDefault();
        }
        else {
            return;
        }
    }
}
CmEnrouteAttr.decorators = [
    { type: Directive, args: [{
                selector: '[cmEnroute]'
            },] }
];
/** @nocollapse */
CmEnrouteAttr.ctorParameters = () => [
    { type: ElementRef },
    { type: Router }
];
CmEnrouteAttr.propDecorators = {
    baseUrl: [{ type: Input, args: ['cmEnroute',] }],
    onClick: [{ type: HostListener, args: ['click', ['$event'],] }]
};
if (false) {
    /** @type {?} */
    CmEnrouteAttr.prototype.baseUrl;
    /**
     * @type {?}
     * @private
     */
    CmEnrouteAttr.prototype.el;
    /**
     * @type {?}
     * @private
     */
    CmEnrouteAttr.prototype.router;
}
