// @chakray/marked/esm2015/marked/marked.tag.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Input, Component, Inject } from '@angular/core';
import { DomSanitizer } from '@angular/platform-browser';
import { MdLib, mdLib } from './lib';
/**
 * <cm-marked str='*markdown*'>
 */
export class CmMarkedTag {
    /**
     * @param {?} sanitizer see external DomSanitizer
     * @param {?} lib see [[MdLib]]
     */
    constructor(sanitizer, lib) {
        this.sanitizer = sanitizer;
        this.lib = lib;
        this.baseUrl = '';
        this._str = '';
    }
    /**
     * str for set markdown string
     * @param {?} v
     * @return {?}
     */
    set str(v) {
        this._str = v;
        /** @type {?} */
        const r = this.lib.parse(v);
        this.html = this.sanitizer.bypassSecurityTrustHtml(r);
    }
}
CmMarkedTag.decorators = [
    { type: Component, args: [{
                selector: 'cm-marked',
                template: "<div [cmEnroute]='baseUrl' [innerHtml]='html'></div>\n",
                styles: [":host{outline:0;border:0}"]
            }] }
];
/** @nocollapse */
CmMarkedTag.ctorParameters = () => [
    { type: DomSanitizer },
    { type: MdLib, decorators: [{ type: Inject, args: [mdLib,] }] }
];
CmMarkedTag.propDecorators = {
    baseUrl: [{ type: Input }],
    str: [{ type: Input }]
};
if (false) {
    /** @type {?} */
    CmMarkedTag.prototype.baseUrl;
    /** @type {?} */
    CmMarkedTag.prototype.html;
    /**
     * @type {?}
     * @private
     */
    CmMarkedTag.prototype._str;
    /**
     * @type {?}
     * @private
     */
    CmMarkedTag.prototype.sanitizer;
    /**
     * @type {?}
     * @private
     */
    CmMarkedTag.prototype.lib;
}
